Iter8 controller crashes

**Describe the bug**
Observed the controller crashing (and restarting) when deleting watched objects.

**To Reproduce**
Watch logs (via `kubectl logs -f` while execting the tutorials in development (https://github.com/iter8-tools/docs/pull/79). When deleting objects, see panic similar to:

```
E0504 17:55:38.854402       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 54 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x23cfee0?, 0x412bc30})
	k8s.io/apimachinery@v0.26.3/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2bdebc8?})
	k8s.io/apimachinery@v0.26.3/pkg/util/runtime/runtime.go:49 +0x75
panic({0x23cfee0, 0x412bc30})
	runtime/panic.go:884 +0x212
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetAPIVersion(...)
	k8s.io/apimachinery@v0.26.3/pkg/apis/meta/v1/unstructured/unstructured.go:218
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GroupVersionKind(0x0)
	k8s.io/apimachinery@v0.26.3/pkg/apis/meta/v1/unstructured/unstructured.go:426 +0x5b
k8s.io/client-go/tools/reference.GetReference(0x2934280?, {0x2bdee98, 0x0})
	k8s.io/client-go@v0.26.3/tools/reference/ref.go:59 +0x142
k8s.io/client-go/tools/record.(*recorderImpl).generateEvent(0xc00063e980, {0x2bdee98?, 0x0}, 0x0?, {0x27df524, 0x6}, {0x2806d47, 0x17}, {0x282ac25, 0x24})
	k8s.io/client-go@v0.26.3/tools/record/event.go:333 +0x79
k8s.io/client-go/tools/record.(*recorderImpl).Event(0xc00079fec0?, {0x2bdee98?, 0x0?}, {0x27df524?, 0x0?}, {0x2806d47?, 0xc0007a60e0?}, {0x282ac25?, 0xc0007a60e0?})
	k8s.io/client-go@v0.26.3/tools/record/event.go:363 +0x4f
github.com/iter8-tools/iter8/controllers.broadcastEvent({0x2bdee98, 0x0}, {0x27df524, 0x6}, {0x2806d47, 0x17}, {0x282ac25, 0x24}, {0x2c16ca8, 0xc00078d5f0})
	github.com/iter8-tools/iter8/controllers/events.go:23 +0x1de
github.com/iter8-tools/iter8/controllers.removeFinalizer.func1()
	github.com/iter8-tools/iter8/controllers/finalizer.go:169 +0x8f2
k8s.io/client-go/util/retry.OnError.func1()
	k8s.io/client-go@v0.26.3/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x24af8c0, 0x7f0bf2e95701})
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:222 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x2bf4d40?, 0xc000132008?}, 0x50?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:235 +0x57
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x25a59e0?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:228 +0x39
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x40dc67?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:423 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x2933c90, 0xc00011c5a0)
	k8s.io/client-go@v0.26.3/util/retry/util.go:50 +0xf1
k8s.io/client-go/util/retry.RetryOnConflict(...)
	k8s.io/client-go@v0.26.3/util/retry/util.go:104
github.com/iter8-tools/iter8/controllers.removeFinalizer({0xc00097a3b8?, 0x8?}, {0xc000a68570?, 0x11?}, {0xc00097aa1c?, 0x4?}, {0x2c16ca8?, 0xc00078d5f0?}, 0xc000119d00?)
	github.com/iter8-tools/iter8/controllers/finalizer.go:109 +0x1ff
github.com/iter8-tools/iter8/controllers.initAppResourceInformers.func2({0x27c2e60?, 0xc00066c010}, {0xc00097aa1c, 0x4}, {0x27e0a6c, 0x6})
	github.com/iter8-tools/iter8/controllers/allcontrollers.go:90 +0x657
github.com/iter8-tools/iter8/controllers.initAppResourceInformers.func4({0x406878?, 0xc000a881e0?}, {0x27c2e60?, 0xc00066c010?})
	github.com/iter8-tools/iter8/controllers/allcontrollers.go:107 +0x49
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	k8s.io/client-go@v0.26.3/tools/cache/controller.go:239
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	k8s.io/client-go@v0.26.3/tools/cache/shared_informer.go:909 +0xf7
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000095790?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x46e147?, {0x2bceda0, 0xc00064c420}, 0x1, 0xc0006521e0)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x201abf8?, 0x3b9aca00, 0x0, 0x15?, 0xc000095788?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:92
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000312200?)
	k8s.io/client-go@v0.26.3/tools/cache/shared_informer.go:905 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:75 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:73 +0x85
E0504 17:55:38.854574       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 54 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x23cfee0?, 0x412bc30})
	k8s.io/apimachinery@v0.26.3/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000670fa8?})
	k8s.io/apimachinery@v0.26.3/pkg/util/runtime/runtime.go:49 +0x75
panic({0x23cfee0, 0x412bc30})
	runtime/panic.go:884 +0x212
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2bdebc8?})
	k8s.io/apimachinery@v0.26.3/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x23cfee0, 0x412bc30})
	runtime/panic.go:884 +0x212
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetAPIVersion(...)
	k8s.io/apimachinery@v0.26.3/pkg/apis/meta/v1/unstructured/unstructured.go:218
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GroupVersionKind(0x0)
	k8s.io/apimachinery@v0.26.3/pkg/apis/meta/v1/unstructured/unstructured.go:426 +0x5b
k8s.io/client-go/tools/reference.GetReference(0x2934280?, {0x2bdee98, 0x0})
	k8s.io/client-go@v0.26.3/tools/reference/ref.go:59 +0x142
k8s.io/client-go/tools/record.(*recorderImpl).generateEvent(0xc00063e980, {0x2bdee98?, 0x0}, 0x0?, {0x27df524, 0x6}, {0x2806d47, 0x17}, {0x282ac25, 0x24})
	k8s.io/client-go@v0.26.3/tools/record/event.go:333 +0x79
k8s.io/client-go/tools/record.(*recorderImpl).Event(0xc00079fec0?, {0x2bdee98?, 0x0?}, {0x27df524?, 0x0?}, {0x2806d47?, 0xc0007a60e0?}, {0x282ac25?, 0xc0007a60e0?})
	k8s.io/client-go@v0.26.3/tools/record/event.go:363 +0x4f
github.com/iter8-tools/iter8/controllers.broadcastEvent({0x2bdee98, 0x0}, {0x27df524, 0x6}, {0x2806d47, 0x17}, {0x282ac25, 0x24}, {0x2c16ca8, 0xc00078d5f0})
	github.com/iter8-tools/iter8/controllers/events.go:23 +0x1de
github.com/iter8-tools/iter8/controllers.removeFinalizer.func1()
	github.com/iter8-tools/iter8/controllers/finalizer.go:169 +0x8f2
k8s.io/client-go/util/retry.OnError.func1()
	k8s.io/client-go@v0.26.3/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x24af8c0, 0x7f0bf2e95701})
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:222 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x2bf4d40?, 0xc000132008?}, 0x50?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:235 +0x57
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x25a59e0?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:228 +0x39
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x40dc67?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:423 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x2933c90, 0xc00011c5a0)
	k8s.io/client-go@v0.26.3/util/retry/util.go:50 +0xf1
k8s.io/client-go/util/retry.RetryOnConflict(...)
	k8s.io/client-go@v0.26.3/util/retry/util.go:104
github.com/iter8-tools/iter8/controllers.removeFinalizer({0xc00097a3b8?, 0x8?}, {0xc000a68570?, 0x11?}, {0xc00097aa1c?, 0x4?}, {0x2c16ca8?, 0xc00078d5f0?}, 0xc000119d00?)
	github.com/iter8-tools/iter8/controllers/finalizer.go:109 +0x1ff
github.com/iter8-tools/iter8/controllers.initAppResourceInformers.func2({0x27c2e60?, 0xc00066c010}, {0xc00097aa1c, 0x4}, {0x27e0a6c, 0x6})
	github.com/iter8-tools/iter8/controllers/allcontrollers.go:90 +0x657
github.com/iter8-tools/iter8/controllers.initAppResourceInformers.func4({0x406878?, 0xc000a881e0?}, {0x27c2e60?, 0xc00066c010?})
	github.com/iter8-tools/iter8/controllers/allcontrollers.go:107 +0x49
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	k8s.io/client-go@v0.26.3/tools/cache/controller.go:239
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	k8s.io/client-go@v0.26.3/tools/cache/shared_informer.go:909 +0xf7
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000095790?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x46e147?, {0x2bceda0, 0xc00064c420}, 0x1, 0xc0006521e0)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x201abf8?, 0x3b9aca00, 0x0, 0x15?, 0xc000095788?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:92
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000312200?)
	k8s.io/client-go@v0.26.3/tools/cache/shared_informer.go:905 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:75 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:73 +0x85
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x9681bb]

goroutine 54 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000670fa8?})
	k8s.io/apimachinery@v0.26.3/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x23cfee0, 0x412bc30})
	runtime/panic.go:884 +0x212
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2bdebc8?})
	k8s.io/apimachinery@v0.26.3/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x23cfee0, 0x412bc30})
	runtime/panic.go:884 +0x212
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetAPIVersion(...)
	k8s.io/apimachinery@v0.26.3/pkg/apis/meta/v1/unstructured/unstructured.go:218
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GroupVersionKind(0x0)
	k8s.io/apimachinery@v0.26.3/pkg/apis/meta/v1/unstructured/unstructured.go:426 +0x5b
k8s.io/client-go/tools/reference.GetReference(0x2934280?, {0x2bdee98, 0x0})
	k8s.io/client-go@v0.26.3/tools/reference/ref.go:59 +0x142
k8s.io/client-go/tools/record.(*recorderImpl).generateEvent(0xc00063e980, {0x2bdee98?, 0x0}, 0x0?, {0x27df524, 0x6}, {0x2806d47, 0x17}, {0x282ac25, 0x24})
	k8s.io/client-go@v0.26.3/tools/record/event.go:333 +0x79
k8s.io/client-go/tools/record.(*recorderImpl).Event(0xc00079fec0?, {0x2bdee98?, 0x0?}, {0x27df524?, 0x0?}, {0x2806d47?, 0xc0007a60e0?}, {0x282ac25?, 0xc0007a60e0?})
	k8s.io/client-go@v0.26.3/tools/record/event.go:363 +0x4f
github.com/iter8-tools/iter8/controllers.broadcastEvent({0x2bdee98, 0x0}, {0x27df524, 0x6}, {0x2806d47, 0x17}, {0x282ac25, 0x24}, {0x2c16ca8, 0xc00078d5f0})
	github.com/iter8-tools/iter8/controllers/events.go:23 +0x1de
github.com/iter8-tools/iter8/controllers.removeFinalizer.func1()
	github.com/iter8-tools/iter8/controllers/finalizer.go:169 +0x8f2
k8s.io/client-go/util/retry.OnError.func1()
	k8s.io/client-go@v0.26.3/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x24af8c0, 0x7f0bf2e95701})
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:222 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x2bf4d40?, 0xc000132008?}, 0x50?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:235 +0x57
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x25a59e0?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:228 +0x39
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x40dc67?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:423 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x2933c90, 0xc00011c5a0)
	k8s.io/client-go@v0.26.3/util/retry/util.go:50 +0xf1
k8s.io/client-go/util/retry.RetryOnConflict(...)
	k8s.io/client-go@v0.26.3/util/retry/util.go:104
github.com/iter8-tools/iter8/controllers.removeFinalizer({0xc00097a3b8?, 0x8?}, {0xc000a68570?, 0x11?}, {0xc00097aa1c?, 0x4?}, {0x2c16ca8?, 0xc00078d5f0?}, 0xc000119d00?)
	github.com/iter8-tools/iter8/controllers/finalizer.go:109 +0x1ff
github.com/iter8-tools/iter8/controllers.initAppResourceInformers.func2({0x27c2e60?, 0xc00066c010}, {0xc00097aa1c, 0x4}, {0x27e0a6c, 0x6})
	github.com/iter8-tools/iter8/controllers/allcontrollers.go:90 +0x657
github.com/iter8-tools/iter8/controllers.initAppResourceInformers.func4({0x406878?, 0xc000a881e0?}, {0x27c2e60?, 0xc00066c010?})
	github.com/iter8-tools/iter8/controllers/allcontrollers.go:107 +0x49
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	k8s.io/client-go@v0.26.3/tools/cache/controller.go:239
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	k8s.io/client-go@v0.26.3/tools/cache/shared_informer.go:909 +0xf7
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000095790?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x46e147?, {0x2bceda0, 0xc00064c420}, 0x1, 0xc0006521e0)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x201abf8?, 0x3b9aca00, 0x0, 0x15?, 0xc000095788?)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:92
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000312200?)
	k8s.io/client-go@v0.26.3/tools/cache/shared_informer.go:905 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:75 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	k8s.io/apimachinery@v0.26.3/pkg/util/wait/wait.go:73 +0x85
```

**Expected behavior**
no panic. Either catch and ignore the nil pointer or do not trigger it at all

**Additional context**
This seems to happen from event recording when finalizers are removed. Probably the object in question is no longer present.

--- a/controllers/finalizer.go
+++ b/controllers/finalizer.go
@@ -157,7 +157,7 @@ func removeFinalizer(name string, namespace string, gvrShort string, client k8sc
 				// broadcast event
 
 				// get resource for event broadcasting
-				r, err := client.Resource(schema.GroupVersionResource{
+				_, err := client.Resource(schema.GroupVersionResource{
 					Group:    config.ResourceTypes[gvrShort].Group,
 					Version:  config.ResourceTypes[gvrShort].Version,
 					Resource: config.ResourceTypes[gvrShort].Resource,
@@ -166,7 +166,8 @@ func removeFinalizer(name string, namespace string, gvrShort string, client k8sc
 					log.Logger.Warnf("could not get resource with name %s in namespace %s", name, namespace)
 				}
 
-				broadcastEvent(r, corev1.EventTypeNormal, "Deleted Iter8 finalizer", "Deleted Iter8 finalizer for resource", client)
+				// removed because of panic event (see https://github.com/iter8-tools/iter8/issues/1455)
+				// broadcastEvent(r, corev1.EventTypeNormal, "Deleted Iter8 finalizer", "Deleted Iter8 finalizer for resource", client)
 			}
 		}
 

--- a/controllers/finalizer.go
+++ b/controllers/finalizer.go
@@ -153,18 +153,6 @@ func removeFinalizer(name string, namespace string, gvrShort string, client k8sc
 				if kubeerrors.IsNotFound(e) {
 					return nil
 				}
-			} else {
-				// broadcast event
-
-				// get resource for event broadcasting
-				_, err := client.Resource(schema.GroupVersionResource{
-					Group:    config.ResourceTypes[gvrShort].Group,
-					Version:  config.ResourceTypes[gvrShort].Version,
-					Resource: config.ResourceTypes[gvrShort].Resource,
-				}).Namespace(namespace).Get(context.TODO(), name, metav1.GetOptions{})
-				if err != nil {
-					log.Logger.Warnf("could not get resource with name %s in namespace %s", name, namespace)
-				}
 			}
 		}
 

--- a/controllers/finalizer.go
+++ b/controllers/finalizer.go
@@ -165,9 +165,6 @@ func removeFinalizer(name string, namespace string, gvrShort string, client k8sc
 				if err != nil {
 					log.Logger.Warnf("could not get resource with name %s in namespace %s", name, namespace)
 				}
-
-				// removed because of panic event (see https://github.com/iter8-tools/iter8/issues/1455)
-				// broadcastEvent(r, corev1.EventTypeNormal, "Deleted Iter8 finalizer", "Deleted Iter8 finalizer for resource", client)
 			}
 		}
 

